### ISSUE TITLE
Floating point exception in TPad

### ISSUE THREAD

**Issue by andriish** [root-project/root#6374]
## Describe the bug
Crash.

```
user@lxplus733 ~]$  g++ pad.cxx $(root-config --glibs --cflags --libs) -o pad  -g
[user@lxplus733 ~]$ ./pad 
Warning in <UnknownClass::SetDisplay>: DISPLAY not set, setting it to pcatlas18.mppmu.mpg.de:0.0

 *** Break *** floating point exception



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
    gdb.printing.register_pretty_printer(gdb.current_objfile(),
    gdb.printing.register_pretty_printer(gdb.current_objfile(),
#0  0x00007f73109b046c in waitpid () from /lib64/libc.so.6
#1  0x00007f731092df62 in do_system () from /lib64/libc.so.6
#2  0x00007f731533f5dc in TUnixSystem::StackTrace() () from /usr/lib64/root/libCore.so.6.22
#3  0x00007f731534206a in TUnixSystem::DispatchSignals(ESignals) () from /usr/lib64/root/libCore.so.6.22
#4  <signal handler called>
#5  0x00007f7313841eee in TPad::ResizePad(char const*) () from /usr/lib64/root/libGpad.so.6.22
#6  0x00007f731384743a in TPad::TPad(char const*, char const*, double, double, double, double, short, short, short) () from /usr/lib64/root/libGpad.so.6.22
#7  0x0000000000400c6d in main (argc=1, argv=0x7ffd71ee3b98) at pad.cxx:9
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum http://root.cern.ch/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#5  0x00007f7313841eee in TPad::ResizePad(char const*) () from /usr/lib64/root/libGpad.so.6.22
#6  0x00007f731384743a in TPad::TPad(char const*, char const*, double, double, double, double, short, short, short) () from /usr/lib64/root/libGpad.so.6.22
#7  0x0000000000400c6d in main (argc=1, argv=0x7ffd71ee3b98) at pad.cxx:9
===========================================================


[user@lxplus733 ~]$ 
```

## Expected behavior
No FPE.

## To Reproduce
Code
```
#include <TCanvas.h>
#include <TPad.h>
#include <fenv.h>

int main(int argc, char **argv) {
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    TCanvas* C= new TCanvas("canvas","canvas",1024,768);
    C->cd();
    TPad* c= new TPad("my","my",0,1,0,1);
    delete c;
    delete C;
    return 0;
}
```
compile
```
g++ pad.cxx $(root-config --glibs --cflags --libs) -o pad  -g
```
run

```
./pad
```

## Setup
ROOT 6.22.02  and gcc as installed on lxplus machines.

## Additional context
A piece of #6344

**Comment by couet** [comment]
it does not comply on Mac:
```
% clang++ pad.cxx $(root-config --glibs --cflags --libs) -o pad
pad.cxx:6:5: error: use of undeclared identifier 'feenableexcept'; did you mean 'feraiseexcept'?
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    ^~~~~~~~~~~~~~
    feraiseexcept
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/fenv.h:299:12: note: 'feraiseexcept' declared here
extern int feraiseexcept(int /* excepts */);
           ^
1 error generated.

```

**Comment by andriish** [comment]
Hi @couet ,

yes, sure.   `feenableexcept`  is not in  `<fenv.h>` on Mac.  One can add some code to have exceptions handling, but let us keep it simple. Linux is used much wider anyway.

Best regards,

Andrii

**Comment by couet** [comment]
I get the crash à Ubuntu:
```
$ g++ pad.cxx $(root-config --glibs --cflags --libs) -o pad  -g
$ ./pad

 *** Break *** floating point exception



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f12a2167c2a in __GI___wait4 (pid=1802950, stat_loc=stat_loc
entry=0x7ffc57fbd1e8, options=options
entry=0, usage=usage
entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
#1  0x00007f12a2167beb in __GI___waitpid (pid=<optimized out>, stat_loc=stat_loc
entry=0x7ffc57fbd1e8, options=options
entry=0) at waitpid.c:38
#2  0x00007f12a20d70e7 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:172
#3  0x00007f12a29bb82e in TUnixSystem::StackTrace() () from /home/sftnight/couet/build/lib/libCore.so
#4  0x00007f12a29b86b5 in TUnixSystem::DispatchSignals(ESignals) () from /home/sftnight/couet/build/lib/libCore.so
#5  <signal handler called>
#6  0x00007f12a26609d0 in TPad::ResizePad(char const*) () from /home/sftnight/couet/build/lib/libGpad.so
#7  0x00007f12a26596e5 in TPad::TPad(char const*, char const*, double, double, double, double, short, short, short) () from /home/sftnight/couet/build/lib/libGpad.so
#8  0x000056441c6732cd in main (argc=1, argv=0x7ffc57fbffc8) at pad.cxx:9
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum http://root.cern.ch/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#6  0x00007f12a26609d0 in TPad::ResizePad(char const*) () from /home/sftnight/couet/build/lib/libGpad.so
#7  0x00007f12a26596e5 in TPad::TPad(char const*, char const*, double, double, double, double, short, short, short) () from /home/sftnight/couet/build/lib/libGpad.so
#8  0x000056441c6732cd in main (argc=1, argv=0x7ffc57fbffc8) at pad.cxx:9
===========================================================


```

**Comment by couet** [comment]
This PR should fix it: https://github.com/root-project/root/pull/6396

**Comment by couet** [comment]
After this PR, the problem can still be reproduced with:
```
#include <TCanvas.h>
#include <TPad.h>
#include <fenv.h>

int main(int argc, char **argv) {
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    TCanvas* C= new TCanvas("canvas","canvas",1024,768);
    C->cd();
    TPad* c= new TPad("my","my",0,0,1,1);
    c->SetPad(0,1,0,1);
    delete c;
    delete C;
    return 0;
}
```
`SetPad` also needs a protection.

**Comment by andriish** [comment]
Hi @couet ,

Well, the classical method of protection is to chack the numbers before the unsafe operation is executed, i.e.

Just before the difisions in Resize method.

Best regards,

Andrii

**Comment by couet** [comment]
nope ... I think much before ... we should not allow a such pad to exist. I just push a new patch in `SetPad`. I have checked that the pad position cannot be changed elsewhere. So it should be fine. Unless you find an other way to set wrong pad position.  Now I will look if ResizePad should be also protected... just in case.

**Comment by andriish** [comment]
Hm... drawing should not be allowed.
Not sure about existing. 

>an other way to set wrong pad position. 

There might be inherited methods, no?

Best regards,

Andrii

**Comment by couet** [comment]
I do not see any use case where you would like to create an invalid Pad. If you look at the initial version of TPad ctor the protection was already there against invalid values (not in the [0,1] range) for the x et y positions. And the pad  was made "zombie" if these values were not correct. The extra check I added is in the same line. Because the x and y might be in range but w and h might be wrong nevertheless. It is a good check to have. And also the equivalent protection in SetPar was missing. I did not see other place where such protection should be added.
